### ISSUE TITLE
LocalJumpError: break from proc-closure on typo in debug prompt

### ISSUE THREAD

**Issue by zenspider** [ruby/debug#1048]
**Your environment**

* `ruby -v`: ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [arm64-darwin22]
* `rdbg -v`: rdbg 1.9.0

**Describe the bug**

In a minitest run, any typo winds up throwing a LocalJumpError and blows out the debugger session. This does NOT happen at a pry-byebug prompt, so this is unique to the debug gem for some reason.

**To Reproduce**

I don't have this isolated... It doesn't reproduce in a simple program... but basically any typo an the debugger prompt:

```
=>#0	block in <main> (2 levels) at test/my_test.rb:35
  #1	[C] BasicObject#instance_exec at test/minitest_should.rb:30
  # and 25 frames (use `bt' command for all frames)
(rdbg) typo
#<NameError: undefined local variable or method `typo' for #<Minitest::Rails::SpecTests::TestBlahBlahBlah...
```

**Expected behavior**

Typos shouldn't blow up.

**Additional context**

Here's a VERY trimmed/santiized version of the output. This run was in debug 1.8.0 but it reproduces in 1.9.0... line numbers might not be right tho.

```
9989 % time bundle exec ruby -Itest test/my_test.rb -n /wtf/
Run options: -n /wtf/ --seed 20440

# Running:

[29, 38] in test/my_test.rb
    30| 
    31|     innocuous_setup1
    32|     innocuous_setup2
    33| 
=>  34| require "debug"; binding.b
    35|     code_I_want_to_debug
    36|   end
    37| 
    38|   it 'should do the thing' do
=>#0	block in <main> (2 levels) at test/my_test.rb:34
  #1	[C] BasicObject#instance_exec at test/minitest_should.rb:30
  # and 25 frames (use `bt' command for all frames)
(ruby) what_the_fuck                                    # (any typo really)
#<NameError: undefined local variable or method `what_the_fuck' for #<Minitest::Rails::SpecTests:: ... ZOMG 1M of crap! ...>

what_the_fuck
^^^^^^^^^^^^^>
#<LocalJumpError: break from proc-closure>
@@@ #<Thread:0x00000001043d8d58 run>
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:1247:in `backtrace'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:1247:in `block in wait_next_action_'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:1245:in `each'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:1245:in `rescue in wait_next_action_'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:1240:in `wait_next_action_'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:866:in `wait_next_action'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:321:in `suspend'
 > GEMS/debug-1.8.0/lib/debug/thread_client.rb:252:in `on_breakpoint'
 > GEMS/debug-1.8.0/lib/debug/breakpoint.rb:69:in `suspend'
 > GEMS/debug-1.8.0/lib/debug/breakpoint.rb:170:in `block in setup'
 > GEMS/debug-1.8.0/lib/debug/session.rb:2616:in `debugger'
 > test/my_test.rb:34:in `block (2 levels) in <main>'
 > PROJ/test/minitest_should.rb:30:in `instance_exec'
 > PROJ/test/minitest_should.rb:30:in `block (2 levels) in before'
 > GEMS/minitest-5.18.0/lib/minitest/test.rb:99:in `block (4 levels) in run'
... minitest ...
 > GEMS/minitest-5.18.0/lib/minitest.rb:83:in `block in autorun'
@@@ #<Thread:0x000000010801de58@DEBUGGER__::SESSION@server GEMS/debug-1.8.0/lib/debug/session.rb:179 sleep_forever>
 > GEMS/debug-1.8.0/lib/debug/session.rb:244:in `pop'
 > GEMS/debug-1.8.0/lib/debug/session.rb:244:in `pop_event'
 > GEMS/debug-1.8.0/lib/debug/session.rb:248:in `session_server_main'
 > GEMS/debug-1.8.0/lib/debug/session.rb:207:in `block in activate'
@@@ #<Thread:0x000000010ed02f50 GEMS/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:323 sleep>
 > GEMS/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:329:in `sleep'
 > GEMS/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:329:in `block in spawn_thread'
@@@ #<Thread:0x000000011040b5a8 GEMS/minitest-5.18.0/lib/minitest/parallel.rb:28 sleep_forever>
 > GEMS/minitest-5.18.0/lib/minitest/parallel.rb:30:in `pop'
 > GEMS/minitest-5.18.0/lib/minitest/parallel.rb:30:in `block (2 levels) in start'
@@@ #<Thread:0x0000000110413ac8 GEMS/minitest-5.18.0/lib/minitest/parallel.rb:28 sleep_forever>
... minitest parallel ...
 > GEMS/minitest-5.18.0/lib/minitest/parallel.rb:30:in `block (2 levels) in start'
@@@ #<Thread:0x0000000110510f48@Timeout stdlib thread GEMS/timeout-0.3.2/lib/timeout.rb:101 sleep>
 > GEMS/timeout-0.3.2/lib/timeout.rb:113:in `sleep'
... timeout ...
 > GEMS/timeout-0.3.2/lib/timeout.rb:111:in `block in create_timeout_thread'
["DEBUGGER Exception: GEMS/debug-1.8.0/lib/debug/thread_client.rb:1251", #<LocalJumpError: break from proc-closure>, ["GEMS/debug-1.8.0/lib/debug/thread_client.rb:458:in `block in frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:457:in `each'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:457:in `rescue in frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:436:in `frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:1043:in `wait_next_action_'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:866:in `wait_next_action'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:321:in `suspend'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:252:in `on_breakpoint'", "GEMS/debug-1.8.0/lib/debug/breakpoint.rb:69:in `suspend'", "GEMS/debug-1.8.0/lib/debug/breakpoint.rb:170:in `block in setup'", "GEMS/debug-1.8.0/lib/debug/session.rb:2616:in `debugger'", "test/my_test.rb:34:in `block (2 levels) in <main>'", "PROJ/test/minitest_should.rb:30:in `instance_exec'", ... minitest stack ...]]
E

Error:
MyTest::BlahBlahBlah#test_0001_fails wtf:
LocalJumpError: break from proc-closure
    test/my_test.rb:34:in `block (2 levels) in <main>'
```

Please let me know if there's any additional information/debugging I can provide.

**Comment by ko1** [comment]
I also couldn't reproduce the issue.

```
[master]$ ruby target.rb
Run options: --seed 18748

# Running:

[1, 10] in target.rb
     1| require 'minitest/autorun'
     2| require 'debug'
     3|
     4| class FooTest < Minitest::Test
     5|   def test_foo
=>   6|     debugger
     7|     assert_equal 'foo', 'foo'
     8|   end
     9| end
    10|
=>#0    FooTest#test_foo at target.rb:6
  #1    block in run (3 levels) at ~/ruby/install/master/lib/ruby/gems/3.3.0+0/gems/minitest-5.20.0/lib/minitest/test.rb:94
  # and 20 frames (use `bt' command for all frames)
(rdbg) p x    # command
eval error: undefined local variable or method `x' for #<FooTest:0x00007fdea5f42cf8>
  (rdbg)/target.rb:1:in `test_foo'
=> nil
(rdbg) x
eval error: undefined local variable or method `x' for #<FooTest:0x00007fdea5f42cf8>
  (rdbg)/target.rb:1:in `test_foo'
nil
```


```
 > GEMS/timeout-0.3.2/lib/timeout.rb:113:in `sleep'
... timeout ...
 > GEMS/timeout-0.3.2/lib/timeout.rb:111:in `block in create_timeout_thread'
["DEBUGGER Exception: GEMS/debug-1.8.0/lib/debug/thread_client.rb:1251", #<LocalJumpError: break from proc-closure>, ["GEMS/debug-1.8.0/lib/debug/thread_client.rb:458:in `block in frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:457:in `each'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:457:in `rescue in frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:436:in `frame_eval'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:1043:in `wait_next_action_'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:866:in `wait_next_action'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:321:in `suspend'", "GEMS/debug-1.8.0/lib/debug/thread_client.rb:252:in `on_breakpoint'", "GEMS/debug-1.8.0/lib/debug/breakpoint.rb:69:in `suspend'", "GEMS/debug-1.8.0/lib/debug/breakpoint.rb:170:in `block in setup'", "GEMS/debug-1.8.0/lib/debug/session.rb:2616:in `debugger'", "test/my_test.rb:34:in `block (2 levels) in <main>'", "PROJ/test/minitest_should.rb:30:in `instance_exec'", ... minitest stack ...]]
```
Does it multi-threaded program with timeout?

If you can make smaller repro-code, it will help a lot.

**Comment by KJTsanaktsidis** [comment]
OK, I was able to reproduce this (in the same codebase as @zenspider) and I think this is a manifestation of https://bugs.ruby-lang.org/issues/18991. If code coverage collection (simplecov) is disabled, then the issue does not occur.

It also is fixed with Ruby 3.2.2, and with the tip of the ruby_3_1 branch (3.1.4-0a70dde90ab2b52191f4782d0ec7d0a99ab1e845), because the fix was backported. However, this isn't fixed in a _released_ version of Ruby 3.1, because there hasn't been a release since the backport was done.
